rkt/api_service.go

```diff
@@ -0,0 +1,559 @@
+	if len(filter.Ids) > 0 {
+		if !findString(image.Id, filter.Ids, stringEqual) {
+			return true
+		}
+	}
+
+	// Filter according to the image name prefixes.
+	if len(filter.Prefixes) > 0 {
+		if !findString(image.Name, filter.Prefixes, strings.HasPrefix) {
+			return true
+		}
+	}
+
+	// Filter according to the image base name.
+	if len(filter.BaseNames) > 0 {
+		if !findString(image.Name, filter.BaseNames, hasBaseName) {
+			return true
+		}
+	}
+
+	// Filter according to the image keywords.
+	if len(filter.Keywords) > 0 {
+		if !findString(image.Name, filter.Keywords, strings.Contains) {
+			return true
+		}
+	}
+
+	// Filter according to the imported time.
+	if filter.ImportedAfter > 0 {
+		if image.ImportTimestamp <= filter.ImportedAfter {
+			return true
+		}
+	}
+	if filter.ImportedBefore > 0 {
+		if image.ImportTimestamp >= filter.ImportedBefore {
+			return true
+		}
+	}
+
+	// Filter according to the image labels.
+	if len(filter.Labels) > 0 {
+		if !findKeyValue(manifest.Labels, filter.Labels) {
+			return true
+		}
+	}
+
+	// Filter according to the annotations.
+	if len(filter.Annotations) > 0 {
+		if !findKeyValue(manifest.Annotations, filter.Annotations) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (s *v1AlphaAPIServer) ListImages(ctx context.Context, request *v1alpha.ListImagesRequest) (*v1alpha.ListImagesResponse, error) {
+	aciInfos, err := s.store.GetAllACIInfos(nil, false)
+	if err != nil {
+		log.Printf("Failed to get all ACI infos: %v", err)
+		return nil, err
+	}
+
+	var images []*v1alpha.Image
+	for _, aciInfo := range aciInfos {
+		image, manifest, err := s.aciInfoToV1AlphaAPIImage(aciInfo)
+		if err != nil {
+			continue
+		}
+
+		// Filter images.
+		if !filterImage(image, manifest, request.Filter) {
+			images = append(images, image)
+		}
+	}
+
+	return &v1alpha.ListImagesResponse{Images: images}, nil
+}
+
+func (s *v1AlphaAPIServer) InspectImage(ctx context.Context, request *v1alpha.InspectImageRequest) (*v1alpha.InspectImageResponse, error) {
+	aciInfo, err := s.store.GetACIInfoWithBlobKey(request.Id)
+	if err != nil {
+		log.Printf("Failed to get ACI info for image ID %q: %v", request.Id, err)
+		return nil, err
+	}
+
+	image, _, err := s.aciInfoToV1AlphaAPIImage(aciInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	return &v1alpha.InspectImageResponse{Image: image}, nil
+}
+
+func (s *v1AlphaAPIServer) GetLogs(request *v1alpha.GetLogsRequest, server v1alpha.PublicAPI_GetLogsServer) error {
+	return fmt.Errorf("not implemented yet")
+}
+
+func (s *v1AlphaAPIServer) ListenEvents(request *v1alpha.ListenEventsRequest, server v1alpha.PublicAPI_ListenEventsServer) error {
+	return fmt.Errorf("not implemented yet")
+}
+
+func runAPIService(cmd *cobra.Command, args []string) (exit int) {
+	log.Print("API service starting...")
+
+	tcpl, err := net.Listen("tcp", flagAPIServiceListenClientURL)
+	if err != nil {
+		log.Printf("Error listening on url %v: %v", flagAPIServiceListenClientURL, err)
```

Comment:
The current behaviour, tested by running `nc -l -p 15441` in the background, is:
```
2015/10/19 18:33:50 Error listening on url localhost:15441: listen tcp 127.0.0.1:15441: bind: address already in use
panic: runtime error: invalid memory address or nil pointer dereference
```

To keep it consistent with other commands and shorten the message:
* `stderr("api-service: %v", err)`
* add `return 1` to avoid the panic.